---
title: "GO-Markdown"
author: "Alexa Huzar, Elsa Brenner, Erica Sun"
date: "April 29, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## GO Enrichment

After running DeSeq and WGCNA, we ran GO enrichment to identify the function of the genes that were differentially expressed as a result of the experimental treatment.  

According to the original GO script that we built this analysis from: "GO uses continuous measure of significance (such as fold-change or -log(p-value) ) to identify GO categories that are significantly enriched with either up- or down-regulated genes. The advantage- no need to impose arbitrary signficance curoffs."  

As you will see in the coming analysis, the result of running GO is a heirarchacal tree that clusters GO categories by shared genes.  

Let's take a look!  

### Running GO 

The first step in any R file is to set your working directory:

```{r}
setwd("/usr4/bi594/ebrenner/ondemand/GitHub-final")
```

Make sure that all of your files from DeSeq and WGCNA are in the same place as you are running GO.  

First, we will be running the GO script from the results of the DeSeq analysis.  We will have to run this script six times-- for each experimental treatment, Heat and Cool, and then for all three GO divisions-- Biological Processes (BP), Molecular Function (MF), and Cellular Componants (CC).  

Here, we will show the fist script in depth, and then point out the places where you would change the R script to get the other figures.  

Let's start with BP for the cool treatment:

You may need to install the package "ape"

```{r}
install.packages("ape")
library("ape")
```

Next, let's load the correct files:

```{r}
input="Cool_GO.csv" # two columns of comma-separated values: gene id, continuous measure of significance. To perform standard GO enrichment analysis based on Fisher's exact test, use binary measure (0 or 1, i.e., either sgnificant or not).
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab" # two-column, tab-delimited, one line per gene, multiple GO terms separated by semicolon. If you have multiple lines per gene, use nrify_GOtable.pl prior to running this script.
goDatabase="go.obo" # download from http://www.geneontology.org/GO.downloads.ontology.shtml
goDivision="BP" # either MF, or BP, or CC
source("gomwu.functions.R")
```

We'll be using this file input template for all the rest of our GO analyses.  To reiterate the annotations in the code block: 
the "input" file is where you input the .csv file from your prior analysis (in this case, the DeSeq), 
the "goAnnotations" file is the annotated transcriptome for the target organism.  We used a file compiled by the Davies Lab for B. psygmospilum.
the "goDatabase" can be downloaded from the url in the code block annotation,
and the "goDivision" is where you specify your disired GO division.  You can see here we are starting with BP.  

The next step is to calculate the stats necessary to create figures.  

```{r}
gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl", # replace with full path to perl executable if it is not in your system's PATH already
           largest=0.1,  # a GO category will not be considered if it contains more than this fraction of the total number of genes
           smallest=5,   # a GO category should contain at least this many genes to be considered
           clusterCutHeight=0.25, # threshold for merging similar (gene-sharing) terms. See README for details.
           #	Alternative="g" # by default the MWU test is two-tailed; specify "g" or "l" of you want to test for "greater" or "less" instead. 
           #	Module=TRUE,Alternative="g" # un-remark this if you are analyzing a SIGNED WGCNA module (values: 0 for not in module genes, kME for in-module genes). In the call to gomwuPlot below, specify absValue=0.001 (count number of "good genes" that fall into the module)
           #	Module=TRUE # un-remark this if you are analyzing an UNSIGNED WGCNA module 
) 
# do not continue if the printout shows that no GO terms pass 10% FDR.
```

the Stats returned 37 GO terms at 10% FDR.  

Since we standardized names in the last code block, there isn't much you will need to change here.  A few things to note are the perlPath, which you will need to make sure correctly matches your system, and the cutoffs.  You can change the "largest" "smallest" and "clusterCutHeight" to your desired values, but for our analyses we will be keeping these standard values.  

This stats block takes a while-- especially for BP-- so if possible run this section through a remote computing cluster (the SCC if you're at BU), or ensure that your local computer can handle the job.  

If you have GO terms, congrats!! you can now plot your results.  If you do not get any GO terms, you can change your cutoffs to encompase a broader range of results, or change the type of analysis you are running.  For this analysis we will run a Man Whitney (MWU) test.    

Next, we're going to plot our results.  

```{r}
results=gomwuPlot(input,goAnnotations,goDivision,
                  #	absValue=-log(0.05,10),  # genes with the measure value exceeding this will be counted as "good genes". Specify absValue=0.001 if you are doing Fisher's exact test for standard GO enrichment or analyzing a WGCNA module (all non-zero genes = "good genes").
                  absValue=1,
                  level1=0.1, # FDR threshold for plotting. Specify level1=1 to plot all GO categories containing genes exceeding the absValue.
                  level2=0.05, # FDR cutoff to print in regular (not italic) font.
                  level3=0.01, # FDR cutoff to print in large bold font.
                  txtsize=1.2,    # decrease to fit more on one page, or increase (after rescaling the plot so the tree fits the text) for better "word cloud" effect
                  treeHeight=0.5, # height of the hierarchical clustering tree
                  #	colors=c("dodgerblue2","firebrick1","skyblue2","lightcoral") # these are default colors, un-remar and change if needed
)
```
Figure 3.1: GO enrichment results for the cool treatment, looking at Biological Processes.  Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  Blue represents downregulation, and red represents upregulation.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.  

You can leave this plotting block unchanged for each analysis! All this does is takes your stats from the last block, and plots it onto a heirarchical tree so that you can visualize your results!  

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
