---
title: "GO-Markdown"
author: "Alexa Huzar, Elsa Brenner, Erica Sun"
date: "April 29, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## GO Enrichment

After running DeSeq and WGCNA, we ran GO enrichment to identify the function of the genes that were differentially expressed as a result of the experimental treatment.  

According to the original GO script that we built this analysis from: "GO uses continuous measure of significance (such as fold-change or -log(p-value) ) to identify GO categories that are significantly enriched with either up- or down-regulated genes. The advantage- no need to impose arbitrary signficance curoffs."  

As you will see in the coming analysis, the result of running GO is a heirarchacal tree that clusters GO categories by shared genes.  

Let's take a look!  

### Running GO 

The first step in any R file is to set your working directory:

```{r}
setwd("/usr4/bi594/ahuzar/R/Symbiont-Thermal-Stress")
```

Make sure that all of your files from DeSeq and WGCNA are in the same place as you are running GO.  

First, we will be running the GO script from the results of the DeSeq analysis.  We will have to run this script six times-- for each experimental treatment, Heat and Cool, and then for all three GO divisions-- Biological Processes (BP), Molecular Function (MF), and Cellular Componants (CC).  

Here, we will show the fist script in depth, and then point out the places where you would change the R script to get the other figures.  

Let's start with BP for the cool treatment:

You may need to install the package "ape"

```{r, echo=FALSE}
library("ape")
```

Next, let's load the correct files:

```{r, message=FALSE, warning=FALSE}
input="Cool_GO.csv" # two columns of comma-separated values: gene id, continuous measure of significance. To perform standard GO enrichment analysis based on Fisher's exact test, use binary measure (0 or 1, i.e., either sgnificant or not).
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab" # two-column, tab-delimited, one line per gene, multiple GO terms separated by semicolon. If you have multiple lines per gene, use nrify_GOtable.pl prior to running this script.
goDatabase="go.obo" # download from http://www.geneontology.org/GO.downloads.ontology.shtml
goDivision="BP" # either MF, or BP, or CC
source("gomwu.functions.R")
```

We'll be using this file input template for all the rest of our GO analyses.  To reiterate the annotations in the code block: 
the "input" file is where you input the .csv file from your prior analysis (in this case, the DeSeq), 
the "goAnnotations" file is the annotated transcriptome for the target organism.  We used a file compiled by the Davies Lab for B. psygmospilum.
the "goDatabase" can be downloaded from the url in the code block annotation,
and the "goDivision" is where you specify your disired GO division.  You can see here we are starting with BP.  

**The next step is to calculate the stats necessary to create figures.**  

```{r, message=FALSE, warning=FALSE}
gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl", # replace with full path to perl executable if it is not in your system's PATH already
           largest=0.1,  # a GO category will not be considered if it contains more than this fraction of the total number of genes
           smallest=5,   # a GO category should contain at least this many genes to be considered
           clusterCutHeight=0.25, # threshold for merging similar (gene-sharing) terms. See README for details.
           #	Alternative="g" # by default the MWU test is two-tailed; specify "g" or "l" of you want to test for "greater" or "less" instead. 
           #	Module=TRUE,Alternative="g" # un-remark this if you are analyzing a SIGNED WGCNA module (values: 0 for not in module genes, kME for in-module genes). In the call to gomwuPlot below, specify absValue=0.001 (count number of "good genes" that fall into the module)
           #	Module=TRUE # un-remark this if you are analyzing an UNSIGNED WGCNA module 
) 
# do not continue if the printout shows that no GO terms pass 10% FDR.
```

the Stats returned 37 GO terms at 10% FDR.  

Since we standardized names in the last code block, there isn't much you will need to change here.  A few things to note are the perlPath, which you will need to make sure correctly matches your system, and the cutoffs.  You can change the "largest" "smallest" and "clusterCutHeight" to your desired values, but for our analyses we will be keeping these standard values.  

This stats block takes a while-- especially for BP-- so if possible run this section through a remote computing cluster (the SCC if you're at BU), or ensure that your local computer can handle the job.  

If you have GO terms, congrats!! you can now plot your results.  If you do not get any GO terms, you can change your cutoffs to encompase a broader range of results, or change the type of analysis you are running.  For this analysis we will run a Man Whitney (MWU) test.    

**Next, we're going to plot our results.**

```{r, message=FALSE, warning=FALSE}
results=gomwuPlot(input,goAnnotations,goDivision,
                  #	absValue=-log(0.05,10),  # genes with the measure value exceeding this will be counted as "good genes". Specify absValue=0.001 if you are doing Fisher's exact test for standard GO enrichment or analyzing a WGCNA module (all non-zero genes = "good genes").
                  absValue=1,
                  level1=0.1, # FDR threshold for plotting. Specify level1=1 to plot all GO categories containing genes exceeding the absValue.
                  level2=0.05, # FDR cutoff to print in regular (not italic) font.
                  level3=0.01, # FDR cutoff to print in large bold font.
                  txtsize=1.2,    # decrease to fit more on one page, or increase (after rescaling the plot so the tree fits the text) for better "word cloud" effect
                  treeHeight=0.5, # height of the hierarchical clustering tree
                  #	colors=c("dodgerblue2","firebrick1","skyblue2","lightcoral") # these are default colors, un-remar and change if needed
)
```

Figure 3.1: GO enrichment results for the cool treatment, looking at Biological Processes.  Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  Blue represents downregulation, and red represents upregulation.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.  

You can leave this plotting block unchanged for each analysis! All this does is takes your stats from the last block, and plots it onto a heirarchical tree so that you can visualize your results!  

## Let's repeat this for all combinations of treatments and GO divisions! 

**CC for cool treatment**

```{r, echo=FALSE, message=FALSE, warning=FALSE}
input="Cool_GO.csv"
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="CC"
source("gomwu.functions.R")


gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl", 
           largest=0.1,  
           smallest=5,   
           clusterCutHeight=0.25,
)
# 32 GO terms at 10% FDR
```
```{r, echo=FALSE, message=FALSE, warning=FALSE}
results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=1,
                  level1=0.1, 
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,    
                  treeHeight=0.5,
)
```

Figure 3.2: GO enrichment results for the cool treatment, looking at Cellular Componants. There are 32 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  Blue represents downregulation, and red represents upregulation.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

**BP for the cool treatment**

```{r, echo=FALSE, message=FALSE, warning=FALSE}
input="Cool_GO.csv" 
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab" 
goDatabase="go.obo" 
goDivision="MF"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl", 
           largest=0.1,  
           smallest=5,  
           clusterCutHeight=0.25,
)
# 44 GO terms at 10% FDR
```
```{r, echo=FALSE, message=FALSE, warning=FALSE}
results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=1,
                  level1=0.1, 
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,    
                  treeHeight=0.5,
)
```

Figure 3.3: GO enrichment results for the cool treatment, looking at Molecular Functions. There are 44 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  Blue represents downregulation, and red represents upregulation.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

### We've finished all GO Divisions for the Cool Treatment, now let's move on to the Heat treatment.  

**BP**

```{r, echo=FALSE, message=FALSE, warning=FALSE}
input="Heat_GO.csv" # Notice we've switched the CSV to the Heat Treatment CSV file.  
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo" 
goDivision="BP"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5, 
           clusterCutHeight=0.25,
)
# 26 GO terms at 10% FDR
```
```{r, echo=FALSE, message=FALSE, warning=FALSE}
results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=1,
                  level1=0.1, 
                  level2=0.05, 
                  level3=0.01, 
                  txtsize=1.2,    
                  treeHeight=0.5, 
)
```

Figure 3.4: GO enrichment results for the heat treatment, looking at Biological Processes. There are 26 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  Blue represents downregulation, and red represents upregulation.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

**CC**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="Heat_GO.csv"
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo" 
goDivision="CC"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1, 
           smallest=5,   
           clusterCutHeight=0.25,
)
# 25 GO terms at 10% FDR
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=1,
                  level1=0.1,
                  level2=0.05, 
                  level3=0.01, 
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.5: GO enrichment results for the heat treatment, looking at Cellular Componants. There are 25 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  Blue represents downregulation, and red represents upregulation.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

**MF**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="Heat_GO.csv" 
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab" 
goDatabase="go.obo"
goDivision="MF"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
)
# 6 GO terms at 10% FDR
```
```{r, echo=FALSE, message=FALSE, warning=FALSE}
results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=1,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.6: GO enrichment results for the heat treatment, looking at Molecular Functions. There are 6 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  Blue represents downregulation, and red represents upregulation.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

## Now we've finished all of our original GO divisions and experimental treatments!  But we're not done with GO...

The above GO analysis uses the information we collected from DeSeq, but we can also apply GO to the outputs of WGCNA analysis.  Looking at the results of WGCNA, we know there were 4 modules of interest: turquoise, greenyellow, magenta, and brown.  We can use a similar code template as before to see the GO functions of the genes within the module. The module are run as Fishers test which simply checks if a GO term is in a module not the direction of expression. 

### Let's take a look at the Turquoise module first: 

Turquoise is positively correlated with the cool treatment

**BP**

```{r, message=FALSE, warning=FALSE}
input="turquoiseturquoise_fisher.csv" # this is the CSV created from WGCNA, the rest of the files are the same as before.  
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab" 
goDatabase="go.obo"
goDivision="BP" # Like before, let's start with biological processes.  
source("gomwu.functions.R")


gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1, 
           smallest=5,  
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
)
#3 GO terms at 10% FDR
```
```{r, message=FALSE, warning=FALSE}
results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.7: GO enrichment results for the Turquoise module, looking at Biological Processes. There are 3 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

#### Now just like before, we're going to have to run each module from WGCNA through all three GO divisions.  

**Turquoise CC**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="turquoiseturquoise_fisher.csv"
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="CC" 
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) # 9 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.9: GO enrichment results for the Turquoise module, looking at Cellular Componants. There are 9 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.


There were no GO terms at 10% FDR for turquoise MF.

## Now we've finished the turquoise module, but we need to run all GO divisions for the other 3 modules.  

### GreenYellow:
Negatively correlated with heat treatment

**BP**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="greenyellowgreenyellow_fisher.csv"
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="BP"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #10 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.10: GO enrichment results for the GreenYellow module, looking at BP. There are 10 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

**MF**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="greenyellowgreenyellow_fisher.csv"
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="MF"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #5 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.11: GO enrichment results for the GreenYellow module, looking at MF. There are 5 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

**CC**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="greenyellowgreenyellow_fisher.csv"
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="CC"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #12 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.12: GO enrichment results for the GreenYellow module, looking at CC. There are 12 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

### Magenta:
Negatively correlated with cool treatment.

**BP**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="magentamagenta_fisher.csv" 
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="BP"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #11 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.13: GO enrichment results for the Magenta module, looking at BP. There are 11 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

**MF**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="magentamagenta_fisher.csv" 
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="MF"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #7 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.14: GO enrichment results for the Magenta module, looking at MF. There are 7 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

**CC**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="magentamagenta_fisher.csv" 
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="CC"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #6 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.15: GO enrichment results for the Magenta module, looking at CC. There are 6 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left of each GO term display the number of enriched terms in the numerator, over the total number of go terms in that unit in the denominator.

### Brown: 
Positively correlated with heat treatment and cell density

Now for Brown, we are going to do things a bit differently.  When we ran the code the first time, using the Fisher test, it produced zero GO terms.  So in order to produce results, we switched to using the kME file produced from WGCNA.  You will see that difference reflected in the code below.  

**BP**

```{r, message=FALSE, warning=FALSE}
input="brownbrown_kME.csv" # here you can see the different file input
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="BP"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #142 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.01,
                  level2=0.001,
                  level3=0.0001,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.16: GO enrichment results for the Brown module, looking at BP. There are 142 GO terms at 10% FDR.Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.001, and <0.0001.  The fractions to the left do not denote anything, and should be ignored for this figure.

**MF**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="brownbrown_kME.csv"
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="MF"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #21 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.01,
                  level2=0.001,
                  level3=0.0001,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.17: GO enrichment results for the Brown module, looking at MF. There are 21 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.001, and <0.001.  The fractions to the left do not denote anything, and should be ignored for this figure.

**CC**

```{r echo=FALSE, message=FALSE, warning=FALSE}
input="brownbrown_kME.csv"
goAnnotations="B_psygmophilum_isogroup_to_GOterm.tab"
goDatabase="go.obo"
goDivision="CC"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25,
           Module=TRUE,Alternative="g"
) #24 GO terms at 10% FDR

results=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=0.001,
                  level1=0.01,
                  level2=0.001,
                  level3=0.0001,
                  txtsize=1.2,
                  treeHeight=0.5,
)
```

Figure 3.18: GO enrichment results for the Brown module, looking at CC. There are 24 GO terms at 10% FDR. Text thickness on the tree signifies goodness of fit, using the p-values <0.01, <0.05, and <0.1.  The fractions to the left do not denote anything, and should be ignored for this figure.


# Conclusion
Based on the GO enrichment of the heat and cold stress samples, different pathways are activated by different kinds of thermal stress. The genes upregulated in the cold stress samples were mainly involved in mRNA processing, RNA splicing, and protein folding. Genes involved in photosynthesis, ion transport, and transmembrane transport were downregulated in cold stress. In the heat stress, many of the upregulated genes were involved in protein and macromolecule localization and transport. The downregulated genes were mainly found in the ribosome and involved in peptide and macromolecule synthesis. 

Using the modules from the WGCNA can give a more detailed picture on what expression is best correlated with the treatments. The turquoise module is perfectly positively correlated with the cool treatment while the magenta module is strongly correlated with it as well but negatively (-0.9). Greenyellow is strongly correlated with downregulation in heat (-0.99) while brown is strongly correlated with upregulation in heat (0.92). Brown is also strongly correlated with cell density (0.97) so GO analysis can provide some insight whether the gene expression is affected by thermal stress or cell density. Similar to the results of GO from Deseq, the turquoise module contained genes mainly involved in mRNA processing and RNA splicing. Magenta module genes were involved with photosynthesis and metabolic/catabolic process. This is not unexpected as thermal stress can disrupt the photosystems. The greenyellow module contained genes mainly involved with the ribosome and peptide and macromolecule synthesis. The brown module is not as clear as kME files were used and they displayed vastly more GO terms than the other modules so the significance is not certain. This module contained photosynthesis related genes as well as of particular interest, response to reactive oxygen species. Based on these results, it appears that the brown module expression is influenced by heat treatment more than cell density as the genes are involved in stress response. The correlation with cell density is likely due to the fact that higher cell density is correlated with heat exposure. Overall, all of our GO enrichment indicates a diverse response to thermal stress that varies by whether it is cold or heat stress.





